Sources/SwiftGRPCNIO/CallHandlers/BidirectionalStreamingCallHandler.swift

```diff
@@ -0,0 +1,42 @@
+import Foundation
+import SwiftProtobuf
+import NIO
+import NIOHTTP1
+
+// Handles bidirectional streaming calls. Forwards incoming messages and end-of-stream events to the observer block.
+// The observer block is implemented by the framework user and calls `context.sendResponse` as needed.
+// To close the call and send the status, fulfill `context.statusPromise`.
+public class BidirectionalStreamingCallHandler<RequestMessage: Message, ResponseMessage: Message>: BaseCallHandler<RequestMessage, ResponseMessage> {
+  public typealias EventObserver = (StreamEvent<RequestMessage>) -> Void
+  fileprivate var eventObserver: EventLoopFuture<EventObserver>?
+  
+  //! FIXME: Do we need to keep the context around at all here?
+  public private(set) var context: StreamingResponseCallContext<ResponseMessage>?
+
+  // We ask for a future of type `EventObserver` to allow the framework user to e.g. asynchronously authenticate a call.
+  // If authentication fails, they can simply fail the observer future, which causes the call to be terminated.
+  public init(channel: Channel, headers: HTTPRequestHead, eventObserverFactory: (StreamingResponseCallContext<ResponseMessage>) -> EventLoopFuture<EventObserver>) {
```

Comment:
Quick question here: the parameter is `headers` but it hands you the whole request. Would recommend to either rename `request:` or to pass `HTTPHeaders` instead of a `HTTPRequestHead` here. You can get the headers using `request.headers`

Reply:
Well spotted! I was originally using `HTTPHeaders` there, then noticed that some other request information (e.g. `.uri`) could be useful to have (e.g. for middlewares). I then forgot to rename the field.

Renamed the field now — please take another look.